sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/resourcehints/ResourceHints.java

```diff
@@ -0,0 +1,270 @@
+
+public class ResourceHints {
+  private static final String MIN_RAM_URN = "beam:resources:min_ram_bytes:v1";
+  private static final String ACCELERATOR_URN = "beam:resources:accelerator:v1";
+
+  // TODO: reference this from a common location in all packages that use this.
+  private static String getUrn(ProtocolMessageEnum value) {
+    return value.getValueDescriptor().getOptions().getExtension(RunnerApi.beamUrn);
+  }
+
+  static {
+    checkState(MIN_RAM_URN.equals(getUrn(StandardResourceHints.Enum.MIN_RAM_BYTES)));
+    checkState(ACCELERATOR_URN.equals(getUrn(StandardResourceHints.Enum.ACCELERATOR)));
+  }
+
+  private static ImmutableMap<String, String> hintNameToUrn =
+      ImmutableMap.<String, String>builder()
+          .put("minRam", MIN_RAM_URN)
+          .put("min_ram", MIN_RAM_URN) // Courtesy alias.
+          .put("accelerator", ACCELERATOR_URN)
+          .build();
+
+  private static ImmutableMap<String, Function<String, ResourceHint>> parsers =
+      ImmutableMap.<String, Function<String, ResourceHint>>builder()
+          .put(MIN_RAM_URN, s -> new BytesHint(BytesHint.parse(s)))
+          .put(ACCELERATOR_URN, s -> new StringHint(s))
+          .build();
+
+  private static ResourceHints empty = new ResourceHints(ImmutableMap.of());
+
+  private final ImmutableMap<String, ResourceHint> hints;
+
+  private ResourceHints(ImmutableMap<String, ResourceHint> hints) {
+    this.hints = hints;
+  }
+
+  public static ResourceHints create() {
+    return empty;
+  }
+
+  public static ResourceHints fromOptions(PipelineOptions options) {
+    ResourceHintsOptions resourceHintsOptions = options.as(ResourceHintsOptions.class);
+    ResourceHints result = create();
+    if (resourceHintsOptions.getResourceHints() == null) {
+      return result;
+    }
+    Splitter splitter = Splitter.on('=').limit(2);
+    for (String hint : resourceHintsOptions.getResourceHints()) {
+      List<String> parts = splitter.splitToList(hint);
+      if (parts.size() != 2) {
+        throw new IllegalArgumentException("Unparsable resource hint: " + hint);
+      }
+      String nameOrUrn = parts.get(0);
+      String stringValue = parts.get(1);
+      String urn;
+      if (hintNameToUrn.containsKey(nameOrUrn)) {
+        urn = hintNameToUrn.get(nameOrUrn);
+      } else if (!nameOrUrn.startsWith("beam:resources:")) {
+        // Allow unknown hints to be passed, but validate a little bit to prevent typos.
+        throw new IllegalArgumentException("Unknown resource hint: " + hint);
+      } else {
+        urn = nameOrUrn;
+      }
+      ResourceHint value = parsers.getOrDefault(urn, s -> new StringHint(s)).apply(stringValue);
+      result = result.withHint(urn, value);
+    }
+    return result;
+  }
+
+  /*package*/ static class BytesHint extends ResourceHint {
+    private static Map<String, Long> suffixes =
+        ImmutableMap.<String, Long>builder()
+            .put("B", 1L)
```

Comment:
Per other comment, should this be case sensitive?

Reply:
Replied on the other comment.